Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m

```diff
@@ -0,0 +1,178 @@
+{
+    [super prepareLayout];
+
+    id<ConversationViewLayoutDelegate> delegate = self.delegate;
+    if (!delegate) {
+        OWSFail(@"%@ Missing delegate", self.tag);
+        [self clearState];
+        return;
+    }
+    if (self.collectionView.bounds.size.width <= 0.f || self.collectionView.bounds.size.height <= 0.f) {
+        OWSFail(@"%@ Collection view has invalid size: %@", self.tag, NSStringFromCGRect(self.collectionView.bounds));
+        [self clearState];
+        return;
+    }
+
+    if (self.hasLayout) {
+        return;
+    }
+    self.hasLayout = YES;
+
+    // TODO: Remove this log statement after we've reduced the invalidation churn.
+    DDLogVerbose(@"%@ prepareLayout", self.tag);
+
+    const int vInset = 15;
+    const int hInset = 10;
+    const int vSpacing = 5;
+    const int viewWidth = (int)floor(self.collectionView.bounds.size.width);
+    const int contentWidth = (int)floor(viewWidth - 2 * hInset);
+
+    NSArray<id<ConversationViewLayoutItem>> *layoutItems = self.delegate.layoutItems;
+
+    CGFloat y = vInset;
+    CGFloat contentBottom = y;
+    BOOL isRTL = self.collectionView.isRTL;
+
+    NSInteger row = 0;
+    for (id<ConversationViewLayoutItem> layoutItem in layoutItems) {
+        CGSize layoutSize = [layoutItem cellSizeForViewWidth:viewWidth contentWidth:contentWidth];
+
+        layoutSize.width = MIN(viewWidth, floor(layoutSize.width));
+        layoutSize.height = floor(layoutSize.height);
+        CGRect itemFrame;
+        switch (layoutItem.layoutAlignment) {
+            case ConversationViewLayoutAlignment_Incoming:
+            case ConversationViewLayoutAlignment_Outgoing: {
+                BOOL isLeft = ((layoutItem.layoutAlignment == ConversationViewLayoutAlignment_Incoming && !isRTL)
+                    || (layoutItem.layoutAlignment == ConversationViewLayoutAlignment_Outgoing && isRTL));
+                if (isLeft) {
+                    itemFrame = CGRectMake(hInset, y, layoutSize.width, layoutSize.height);
+                } else {
+                    itemFrame
+                        = CGRectMake(viewWidth - (hInset + layoutSize.width), y, layoutSize.width, layoutSize.height);
+                }
+                break;
+            }
+            case ConversationViewLayoutAlignment_FullWidth:
+                itemFrame = CGRectMake(0, y, viewWidth, layoutSize.height);
+                break;
+            case ConversationViewLayoutAlignment_Center:
+                itemFrame = CGRectMake(
+                    hInset + round((viewWidth - layoutSize.width) * 0.5f), y, layoutSize.width, layoutSize.height);
+                break;
+        }
+
+        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:row inSection:0];
+        UICollectionViewLayoutAttributes *itemAttributes =
+            [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
+        itemAttributes.frame = itemFrame;
+        self.itemAttributesMap[@(row)] = itemAttributes;
```

Comment:
Yup, there's perf important work coming up.  With JSQ we have to invalidate our layout and/or reload our cells way too often.  We should be able to reduce that.  

Having said that, re-layout should now be much cheaper since most of the cost we in cell measurement which is now cached.